src/emoji/data.js

```diff
@@ -1,14 +1,70 @@
 /* @flow strict-local */
 import * as typeahead from '@zulip/shared/js/typeahead';
 
-import type { EmojiType, ReactionType, EmojiForShared } from '../types';
+import * as api from '../api';
+import * as logging from '../utils/logging';
+import type {
+  ThunkAction,
+  PerAccountAction,
+  ServerEmojiData,
+  EmojiType,
+  ReactionType,
+  EmojiForShared,
+} from '../types';
+import { tryParseUrl } from '../utils/url';
+import { REFRESH_SERVER_EMOJI_DATA } from '../actionConstants';
+import { NetworkError } from '../api/apiErrors';
 import { objectFromEntries } from '../jsBackport';
 import { unicodeCodeByName, override } from './codePointMap';
 import zulipExtraEmojiMap from './zulipExtraEmojiMap';
 import { objectEntries } from '../flowPonyfill';
 
 const unicodeEmojiNames = Object.keys(unicodeCodeByName);
 
+const refreshServerEmojiData = (data: ServerEmojiData): PerAccountAction => ({
+  type: REFRESH_SERVER_EMOJI_DATA,
+  data,
+});
+
+export const maybeRefreshServerEmojiData =
+  (server_emoji_data_url: string | void): ThunkAction<Promise<void>> =>
+  async (dispatch, getState) => {
+    if (server_emoji_data_url === undefined) {
+      // The server is too old to support this feature.
+      // TODO(server-6.0): Simplify away; should always be present.
+      return;
+    }
+    const parsedUrl = tryParseUrl(server_emoji_data_url);
+    if (!parsedUrl) {
+      logging.error('Invalid URL at server_emoji_data_url in /register response');
+      return;
+    }
+
+    let data = undefined;
+    try {
+      data = await api.fetchServerEmojiData(parsedUrl);
+    } catch (errorIllTyped) {
+      const e: mixed = errorIllTyped; // https://github.com/facebook/flow/issues/2470
+
+      if (!(e instanceof Error)) {
+        logging.error('Unexpected non-error thrown from api.fetchServerEmojiData.');
+        return;
+      }
+
+      if (e instanceof NetworkError) {
```

Comment:
One scenario this makes me curious about in the caching behavior: suppose by the time we attempt to fetch emoji data, the network is gone entirely, but the URL is one we've previously fetched and should have cached. Will we get the cached result even though the network isn't available at all?

Ideally we should, because the cache entry should mean the network is irrelevant. If indeed we do, then that would be reassuring about the effectiveness of the cache.

Reply:
I've just tried this on an iOS simulator with "Debug JS Remotely" enabled, and it worked as hoped.

I switched out a button's `onPress` handler so that it instead made the emoji-data request, so I could do it on demand. Then I turned off my computer's WiFi. Other requests, like fetch-messages requests, failed, but this one still succeeded and gave us the data.

Reply:
OK, and just did the same thing on an Android emulator and got the same result.